### PR TITLE
Base Command Class

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
I have created the base command class as discussed. Eventually this class will contain the `logging` set up and logs `fatal` errors. But this will be a latter PR 

It contains the `initialize` method that calls the `setup` function and `run` function. The base class has abstract functions for `setup` and `run` which through an error unless they are overridden. 

As a lot of the commands capture interrupt, their is a `handle_interrupt` method as well. This method by default re-raises the interrupt unless overridden.  It also will prevent interrupt from being logged (when logging is implemented).

### CODE REVIEW

File: metalware-new/src/commands/build.rb [hunk 1]
```diff
@@ -1,6 +1,7 @@
 
 require 'active_support/core_ext/string/strip'
 
+require 'commands/basecommand'
```
Comment (BobWhitelock) [inline metalware-new/src/commands/build.rb]: The usual convention in Ruby is to name the class/module using `CamelCase` (which you've done) and then the file using `underscore_case` - very minor but can we name the file `base_command` :slightly_smiling_face:?

**Comment by WilliamMcCumstie** [comment]
Hold off merging this, I just noticed the `repo` commands

File: metalware-new/src/commands/build.rb [hunk 2]
```diff
@@ -113,7 +112,7 @@ def clear_up_built_node_marker_files
         FileUtils.rm_rf(files)
       end
 
-      def handle_interrupt
+      def handle_interrupt(_not_used)
```
Comment (BobWhitelock) [inline metalware-new/src/commands/build.rb]: I was a bit confused by this parameter for a moment, though I now realise it's so the signature matches that in `BaseCommand`. Could be worth giving it the same name as in `BaseCommand` to make this clear? You can still prefix it with an underscore if you prefer to indicate it's required but intentionally not used.

File: metalware-new/src/commands/hunter.rb [hunk 3]
```diff
@@ -123,7 +122,7 @@ def sequenced_name
         "#{@options.prefix}#{@detection_count.to_s.rjust(@options.length, '0')}"
       end
 
-      def handle_interrupt
+      def handle_interrupt(_not_used)
```
Comment (BobWhitelock) [inline metalware-new/src/commands/hunter.rb]: Similarly to comment on `Build#handle_interrupt`.

**Comment by WilliamMcCumstie** [comment]
Ready to for review

File: metalware-new/src/commands/repo/update.rb [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
 
 require 'constants'
 
 module Metalware
   module Commands
     module Repo
-      class Update
-	def initialize(_args, options)
+      class Update < BaseCommand
+	      def setup(args, options)
```
Comment (BobWhitelock) [inline metalware-new/src/commands/repo/update.rb]: Your indentation looks messed up here.

File: metalware-new/src/commands/repo/update.rb [hunk 5]
```diff
@@ -44,23 +49,23 @@ def initialize(_args, options)
             #                         "been reached!"
             raise "Internal error. Check metal log"
           end
-	end
+	      end
```
Comment (BobWhitelock) [inline metalware-new/src/commands/repo/update.rb]: Indentation also looks shifted here.

File: metalware-new/src/commands/repo/update.rb (hunk 4 above)
Comment (WilliamMcCumstie) [inline metalware-new/src/commands/repo/update.rb]: Yeah the indentation was a bit screw when I opened it. Will have another pass over the files

**Comment by WilliamMcCumstie** [comment]
Fixed

**Comment by BobWhitelock** [comment]
Thanks!

**Pull request merged**